akka-actor/src/main/scala/akka/pattern/RetrySupport.scala

```diff
@@ -30,17 +30,73 @@ trait RetrySupport {
+    retryF(attempt, attempts, delay, 1)((_, _) ⇒ ())
+
+  /**
+   * Given a function from Unit to Future, returns an internally retrying Future
+   * The first attempt will be made immediately, each subsequent attempt will be made after 'delay' * 'backoff'
+   * A scheduler (eg context.system.scheduler) must be provided to delay each retry
+   * If attempts are exhausted the returned future is simply the result of invoking attempt.
+   * Note that the attempt function will be invoked on the given execution context for subsequent
+   * tries and therefore must be thread safe (not touch unsafe mutable state).
+   *
+   * <b>Example usage:</b>
+   *
+   * {{{
+   * protected val sendAndReceive: HttpRequest => Future[HttpResponse]
+   * private val sendReceiveRetry: HttpRequest => Future[HttpResponse] = (req: HttpRequest) => retry[HttpResponse](
+   *   attempt = () => sendAndReceive(req),
+   *   attempts = 10,
+   *   delay = 2 seconds,
+   *   backoff = 2)(
+   *    ec = context.system
+   *    scheduler = context.system.scheduler
+   * )
+   * }}}
+   */
+  def retry[T](attempt: () ⇒ Future[T], attempts: Int, delay: FiniteDuration, backoff: Int)(implicit ec: ExecutionContext, scheduler: Scheduler): Future[T] =
+    retryF(attempt, attempts, delay, backoff)((_, _) ⇒ ())
+
+  /**
+   * Given a function from Unit to Future, returns an internally retrying Future
+   * The first attempt will be made immediately, each subsequent attempt will be made after 'delay' * 'backoff'
+   * The callback function will be executed before each retry.
+   * A scheduler (eg context.system.scheduler) must be provided to delay each retry
+   * If attempts are exhausted the returned future is simply the result of invoking attempt.
+   * Note that the attempt and callback functions will be invoked on the given execution context for subsequent
+   * tries and therefore must be thread safe (not touch unsafe mutable state).
+   *
+   * <b>Example usage:</b>
+   *
+   * {{{
+   * protected val sendAndReceive: HttpRequest => Future[HttpResponse]
+   * private val sendReceiveRetry: HttpRequest => Future[HttpResponse] = (req: HttpRequest) => retry[HttpResponse](
+   *   attempt = () => sendAndReceive(req),
+   *   attempts = 10,
    *   delay = 2 seconds,
-   *   scheduler = context.system.scheduler
+   *   backoff = 2) { case (t, currentRetry) =>
+   *     log.warning(
+   *       s"""
+   *         |Operation failed due to ${t.getMessage}.
+   *         |Will retry $currentRetry times, after ${delay * math.pow(backoff, attempts - currentRetry + 1)}.
+   *       """.stripMargin)
    * )
    * }}}
    */
-  def retry[T](attempt: () ⇒ Future[T], attempts: Int, delay: FiniteDuration)(implicit ec: ExecutionContext, scheduler: Scheduler): Future[T] = {
+  def retryF[T](attempt: () ⇒ Future[T], attempts: Int, delay: FiniteDuration, backoff: Int)(f: (Throwable, Int) ⇒ Unit)(implicit ec: ExecutionContext, scheduler: Scheduler): Future[T] = {
```

Comment:
I think there is no need to rename `retry` to `retryF`

Reply:
It is required... In Scala you can't overload a function with an extra list of parameters :-(

Reply:
Sure... Do you have a suggestion? I saw that `xyzPF` is a convention in akka, but this is not really a partial function. So i only kept the `F` :-)

Reply:
`(Throwable, Int) ⇒ Unit` to `(Throwable, Int) ⇒ NextBackOff)` seems more useful.

Reply:
Instead of having a callback the (Throwable, Int) could be provided as parameter to the `attempt`?

```
attempt: Option[(Throwable, Int)] ⇒ Future[T]
```

or
```
firstAttempt: () ⇒ Future[T],
subsequentAttempt: (Throwable, Int) ⇒ Future[T]
```

Reply:
Yes, that would work...

But they have a major drawback in my view. You force the 'callback' to return the same type of result as the `attempt`. That burdens the client, having to provide some sort of to-be-ignored-default value of type `Future[T]`

Is that what we want?